File: src/main/java/kosukeroku/token_radar/model/Coin.java
package kosukeroku.token_radar.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.Data;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "coins")
@Data
public class Coin {
    @Id
    private String id;
    private String symbol;
    private String name;
    private boolean active = true;
    private LocalDateTime lastUpdated;
    private String imageUrl;
    private Integer marketCapRank;

    // current prices
    @Column(precision = 30, scale = 12)
    private BigDecimal currentPrice;

    private Double priceChange24h;
    private Double priceChangePercentage24h;

    @Column(precision = 30, scale = 2)
    private BigDecimal marketCap;

    @Column(precision = 30, scale = 2)
    private BigDecimal totalVolume;

    // price changes
    private Double priceChangePercentage1h;
    private Double priceChangePercentage7d;
    private Double priceChangePercentage30d;

    @Column(columnDefinition = "TEXT")
    private String sparklineData;

    @Column(precision = 30, scale = 12)
    private BigDecimal high24h;

    @Column(precision = 30, scale = 12)
    private BigDecimal low24h;

    // ATH data
    @Column(precision = 30, scale = 12)
    private BigDecimal ath;

    private Double athChangePercentage;
    private LocalDateTime athDate;

    // ATL data
    @Column(precision = 30, scale = 12)
    private BigDecimal atl;

    private Double atlChangePercentage;
    private LocalDateTime atlDate;

    // supply
    @Column(precision = 30, scale = 2)
    private BigDecimal circulatingSupply;
}

File: frontend/src/types/coin.ts
export interface Coin {
    id: string;
    name: string;
    symbol: string;
    imageUrl: string;
    currentPrice: number;
    priceChangePercentage24h: number;
    marketCapRank: number;
    marketCap: number;
    totalVolume: number;
}

export interface CoinDetail extends Coin {
    priceChange24h: number;
    priceChangePercentage1h: number;
    priceChangePercentage7d: number;
    priceChangePercentage30d: number;
    high24h: number;
    low24h: number;
    ath: number;
    athChangePercentage: number;
    athDate: string;
    atl: number;
    atlChangePercentage: number;
    atlDate: string;
    circulatingSupply: number;
    sparklineData: string;
    lastUpdated: string;
}

export interface CoinPage {
    content: CoinDetail[];
    totalPages: number;
    totalElements: number;
    size: number;
    number: number;
}

export interface TrackedCurrency {
    id: number;
    coinId: string;
    coinName: string;
    coinSymbol: string;
    coinImageUrl: string;
    addedAt: string;
    currentPrice: number;
    priceChangePercentage24h: number;
    marketCapRank: number;
    marketCap: number;
    totalVolume: number;
}

export interface UserProfile {
    id: number;
    username: string;
    email: string;
    createdAt: string;
    trackedCurrencies: TrackedCurrency[];
    trackedCount: number;
}

export interface ApiResponse<T> {
    content: T[];
    totalPages: number;
    totalElements: number;
    size: number;
    number: number;
}

export interface AuthResponse {
    token: string;
    type: string;
    id: number;
    username: string;
    email: string;
}

export interface TrackButtonProps {
    coinId: string;
    coinName?: string;
    size?: "sm" | "default" | "lg" | "icon";
    variant?: "default" | "outline" | "secondary" | "ghost";
    className?: string;
    showText?: boolean;
    tableStyle?: boolean;
    onTrackChange?: (isTracked: boolean) => void;
}